File: src/evidence_extractor.py
"""Evidence Extractor — analyzes interactions to find reinforcing/contradicting evidence."""

import json
from uuid import UUID
from src.models import Evidence, Belief
from src.claude import call_claude_json
from src.prompts import evidence_extraction_prompt


def format_beliefs_for_prompt(beliefs: list[Belief]) -> str:
    lines = []
    for i, b in enumerate(beliefs):
        lines.append(
            f"[{i}] \"{b.content}\" "
            f"(confidence={b.confidence:.2f}, tension={b.tension:.2f}, domain={b.domain})"
        )
    return "\n".join(lines)


async def extract_evidence(
    user_message: str,
    assistant_response: str,
    active_beliefs: list[Belief],
) -> list[Evidence]:
    """Extract evidence from an interaction and match it against beliefs."""
    if not active_beliefs:
        return []

    beliefs_summary = format_beliefs_for_prompt(active_beliefs)
    prompt = evidence_extraction_prompt(user_message, assistant_response, beliefs_summary)

    try:
        raw = await call_claude_json("You are a precise evidence extraction system. Respond only in valid JSON.", prompt)
    except (json.JSONDecodeError, Exception):
        return []

    if not isinstance(raw, list):
        return []

    evidence_list = []
    for item in raw:
        if not isinstance(item, dict):
            continue

        belief_index = item.get("belief_index")
        belief_id = None
        if isinstance(belief_index, int) and 0 <= belief_index < len(active_beliefs):
            belief_id = active_beliefs[belief_index].id

        stance = item.get("stance", "neutral")
        if stance not in ("reinforcing", "contradicting", "neutral"):
            stance = "neutral"

        evidence_list.append(Evidence(
            claim=item.get("claim", ""),
            evidence_type=item.get("type", "factual"),
            relevance=item.get("relevance", ""),
            stance=stance,
            belief_id=belief_id,
            strength=min(1.0, max(0.0, float(item.get("strength", 0.5)))),
        ))

    return evidence_list

File: ts/src/evidence-extractor.ts
/**
 * Evidence Extractor — analyzes interactions to find reinforcing/contradicting evidence.
 *
 * Production features:
 * - Structured logging
 * - Robust error handling (returns empty list on failure, never throws)
 * - Input validation and sanitization
 * - Size limits on extracted evidence
 */

import { callClaudeJson } from "./claude.js";
import { evidenceExtractionPrompt, formatBeliefsForPrompt } from "./prompts.js";
import type { Belief, Evidence } from "./types.js";
import { createLogger } from "./logger.js";

const log = createLogger("evidence");

const MAX_EVIDENCE_ITEMS = 20;
const MAX_CLAIM_LENGTH = 500;

export async function extractEvidence(
  userMessage: string,
  assistantResponse: string,
  activeBeliefs: Belief[],
): Promise<Evidence[]> {
  if (!activeBeliefs.length) {
    log.debug("No active beliefs — skipping evidence extraction");
    return [];
  }

  if (!userMessage || !assistantResponse) {
    log.warn("Missing user message or response — skipping extraction");
    return [];
  }

  const timer = log.time("evidence_extraction");
  const summary = formatBeliefsForPrompt(activeBeliefs);
  const prompt = evidenceExtractionPrompt(userMessage, assistantResponse, summary);

  let raw: any[];
  try {
    raw = await callClaudeJson<any[]>(
      "You are a precise evidence extraction system. Respond only in valid JSON.",
      prompt,
    );
  } catch (err) {
    log.error("Evidence extraction LLM call failed", err);
    return [];
  }

  if (!Array.isArray(raw)) {
    log.warn("Evidence extraction returned non-array", { type: typeof raw });
    return [];
  }

  const result: Evidence[] = [];

  for (const item of raw.slice(0, MAX_EVIDENCE_ITEMS)) {
    if (typeof item !== "object" || !item) continue;

    const beliefIndex = item.belief_index;
    let beliefId: string | null = null;

    if (
      typeof beliefIndex === "number" &&
      beliefIndex >= 0 &&
      beliefIndex < activeBeliefs.length
    ) {
      beliefId = activeBeliefs[beliefIndex].id;
    }

    let stance = item.stance as string;
    if (!["reinforcing", "contradicting", "neutral"].includes(stance)) {
      stance = "neutral";
    }

    const claim = String(item.claim ?? "").slice(0, MAX_CLAIM_LENGTH);
    if (!claim) continue; // Skip empty claims

    let evidenceType = item.type ?? "factual";
    if (!["factual", "feedback", "outcome"].includes(evidenceType)) {
      evidenceType = "factual";
    }

    result.push({
      claim,
      evidence_type: evidenceType as Evidence["evidence_type"],
      relevance: String(item.relevance ?? ""),
      stance: stance as Evidence["stance"],
      belief_id: beliefId,
      strength: Math.min(1, Math.max(0, parseFloat(item.strength ?? "0.5"))),
    });
  }

  timer();
  log.info("Evidence extracted", {
    total: result.length,
    reinforcing: result.filter((e) => e.stance === "reinforcing").length,
    contradicting: result.filter((e) => e.stance === "contradicting").length,
    neutral: result.filter((e) => e.stance === "neutral").length,
  });

  return result;
}
